org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/index/ResourceDescriptionStrategyFragment.xtend

```diff
@@ -0,0 +1,123 @@
+package org.eclipse.xtext.xtext.generator.index
+
+import com.google.inject.Inject
+import org.eclipse.emf.ecore.EClass
+import org.eclipse.emf.ecore.EObject
+import org.eclipse.xtend.lib.annotations.Accessors
+import org.eclipse.xtend2.lib.StringConcatenationClient
+import org.eclipse.xtext.AbstractRule
+import org.eclipse.xtext.resource.IDefaultResourceDescriptionStrategy
+import org.eclipse.xtext.resource.IEObjectDescription
+import org.eclipse.xtext.util.IAcceptor
+import org.eclipse.xtext.xtext.AnnotationNames
+import org.eclipse.xtext.xtext.generator.AbstractXtextGeneratorFragment
+import org.eclipse.xtext.xtext.generator.XtextGeneratorNaming
+import org.eclipse.xtext.xtext.generator.model.FileAccessFactory
+import org.eclipse.xtext.xtext.generator.model.GuiceModuleAccess
+import org.eclipse.xtext.xtext.generator.model.TypeReference
+
+import static extension org.eclipse.xtext.GrammarUtil.*
+import static extension org.eclipse.xtext.xtext.generator.model.TypeReference.*
+
+/**
+ * @author Holger Schill - Initial contribution and API
+ */
+class ResourceDescriptionStrategyFragment extends AbstractXtextGeneratorFragment {
+	@Accessors
+	boolean isGenerateStub = true;
+	
+	@Accessors
+	boolean isGenerateXtendStub;
+	
+	@Inject extension XtextGeneratorNaming
+	@Inject FileAccessFactory fileAccessFactory
+	
+	protected def getResourceDescriptionStrategyClass(){
+		return new TypeReference(grammar.runtimeBasePackage + '.index.' + grammar.simpleName + 'DefaultResourceDescriptionStrategy')
+	}
+	
+	protected def getStubResourceDescriptionStrategyClass(){
+		return new TypeReference(grammar.runtimeBasePackage + '.index.' + grammar.simpleName + 'ResourceDescriptionStrategy')
+	}
+	
+	protected def getDefaultResourceDescriptionSuperClass(){
+		return new TypeReference("org.eclipse.xtext.resource.impl.DefaultResourceDescriptionStrategy");
+	}
+	
+	protected def contributeRuntimeGuiceBindings() {
+		val bindingFactory = new GuiceModuleAccess.BindingFactory
+		if(generateStub || generateXtendStub){
+			bindingFactory.addTypeToType(IDefaultResourceDescriptionStrategy.typeRef(), getStubResourceDescriptionStrategyClass);
+		} else {
+			bindingFactory.addTypeToType(IDefaultResourceDescriptionStrategy.typeRef(), resourceDescriptionStrategyClass);
+		}
+		bindingFactory.contributeTo(language.runtimeGenModule)
+	}
+	
+	protected def getExportedRulesFromGrammar(){
+		val alreadyCollected = newHashSet
+		grammar.rules.filter[isExported && alreadyCollected.add(it.type.classifier)].toList
+	}
+	
+	override generate() {
+		val exportedRules = exportedRulesFromGrammar;
+		if(exportedRules.shouldGenerate) {
+			contributeRuntimeGuiceBindings
+			generateResourceDescriptionStrategy(exportedRules)
+		}
+	}
+	
+	protected def shouldGenerate(Iterable<AbstractRule> exportedRules) {
+		return !exportedRules.empty
+	}
+	
+	protected def StringConcatenationClient generateSuperResourceDescriptionStrategyContent(Iterable<AbstractRule> exportedRules) {
```

Comment:
does this solve the rule vs type problem?

Reply:
No that's done here: val alreadyCollected = newHashSet
		grammar.rules.filter[isExported && alreadyCollected.add(it.type.classifier)].toList

Reply:
yes the user should be able to opt out